Allow user to exclude locations

tanjo3 opened the pull request:
This PR adds an "Excluded Locations" tab that lets users specify progression locations to be ignored by the randomizer.

<img src="https://github.com/user-attachments/assets/d103d094-6155-4bdb-9b10-3e51c342b441" width="600" />

The interface here is similar to selecting starting items from randomized items. On the "Excluded Locations" tab, the user can view all the progression locations for the settings selected from the main tab. They can use the buttons to move locations to the "Excluded Locations" side; locations on this side will be considered non-progress locations by the randomizer. Additionally, two filters are added for both progression and excluded locations, allowing the user to quickly find a location in the list. The filters do not affect the final value for either location list.

A few things to note:

- If you have randomized charts on, all 49 sunken treasure locations will be considered progression locations, since we have yet to determine which chart maps where. If you select sunken treasure locations to be excluded, they will contain non-progress items, and the charts leading to those locations will also be considered non-progress.
- Building on that, items solely required to unlock excluded locations will be considered non-progress. For example, even if Miscellaneous is on, if the Ankle check is excluded, Tingle statues will be regarded as non-progress. This means they will be in wooden chests for the purposes of chest type matches contents.
- Excluded locations are considered nonprogress and are not included in hint generation. For example, excluding Ganon's Maze chest means it will never appear as a location hint.
- If you have dungeons on but exclude too many checks in a dungeon so the randomizer can't place dungeon items, you will get an error.
- This is not meant to be an interface for selecting which bosses are required for Required Bosses Mode.

I've also added zlib compression, alongside base64 encoding, for permalinks. This is optional. However, since each location adds a bit to the bitstring, it can get rather long. For the default settings and a seed name of A:
Before: `MS4xMC4wXzBjMjYyNmQAQQBJEFAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGAAA+wLIJQAAAAAAAQBBAIAA`
After: `eJwz1DM00DOIN0g2MjMyS2FwZPAUCGAgEkgwMPxmOqEKZjMCtTYwAAA+Dgbz`

If any further revisions or clarifications are needed, please let me know.

tanjo3 committed:
Allow user to select specific locations to exclude from progression

tanjo3 committed:
Use `zlib` to compress permalinks further

tanjo3 committed:
Fix encode/decode mismatch with `progression_locations` option

tanjo3 committed:
Ensure logic state is correct before randomizing items

tanjo3 committed:
Add test cases for location exclusion

tanjo3 commented:
I've added 10 tests that hopefully cover a sufficient number of setting combos to test this feature.

wooferzfg commented on test/test_dry.py:
```diff
@@ -41,6 +42,8 @@ def enable_all_progression_location_options(options: Options):
 def enable_all_options(options: Options):
   enable_all_progression_location_options(options)
+  # options.progression_locations
+  # options.excluded_locations
```
Did you mean to leave these comments in?

tanjo3 replied:
Yeah, I figured it was in the style of the test function to have all the options there, but the unmodified ones commented out. Like with `randomized_gear`/`starting_gear`.

wooferzfg commented:
I got an error when running the [Windows build](https://github.com/LagoLunatic/wwrando/actions/runs/9439187875/artifacts/1583442849):

![Wind_Waker_Randomizer_4CZGSuPJp1](https://github.com/LagoLunatic/wwrando/assets/8262173/f6cae9b0-ff6c-42f3-a945-d46dbb7615c4)

LagoLunatic commented:
I fixed the pkg error on master, you can revert the workaround in wwrando.spec

tanjo3 committed:
Fix incorrect counting of excluded charts

tanjo3 committed:
Properly set and count items as nonprogress for excluded locations

tanjo3 committed:
Change UI object names to be more descriptive

tanjo3 committed:
Add UI elements to tabstops

tanjo3 committed:
Fix excluded sunken treasures not being added to `nonprogress_locations`

tanjo3 committed:
Check that excluded sunken treasures are properly non-progress locations

tanjo3 committed:
Avoid setting `progression_locations` unnecessarily

tanjo3 committed:
Remove redundant assignment

tanjo3 committed:
Add search filters for the locations

tanjo3 committed:
Prevent locations from potentially being added to `locations_to_filter_out` twice

tanjo3 committed:
Convert list to set for efficient lookup

tanjo3 committed:
Ensure that excluded locations are always sorted internally

Also validates `excluded_locations` to ensure it only contains real locations. This change ensures consistency when `Options` is constructed directly (such as in tests).

tanjo3 committed:
Show version mismatch dialog when pasting an older permalink

tanjo3 committed:
Properly hide `progression_locations` from permalink

tanjo3 committed:
Move excluded locations to `validate`

Running the randomizer with `-noui` or creating `Options` in tests misses `__post_init__`, so move validation to `validate` which gets called by the randomizer itself. Also, dedup excluded locations during validation.

tanjo3 committed:
Add parameter to return cached item locations dict without deepcopy

Most UI interactions call `update_settings` which will trigger a deepcopy of the item locations dict. This is an unnecessary cost for the purpose of safety, so add a parameter to `load_and_parse_item_locations` that allows it to skip the  deepcopy. It's the responsibility of the caller to ensure the dict is not mutated. The default is to deepcopy and we only use `deepcopy=False` in 3 places where the dict is confirmed to not be mutated.

tanjo3 committed:
Fix circular import
